### PR TITLE
Unboxed float type parsing in `layouts_alpha`

### PR DESCRIPTION
**Pull request by ncik-roberts**
Parse `float#` as the unboxed float type in `layouts_alpha`, and make it alias to `float` for now so I can write some tests.

As discussed, `float# c` and `float#c` are now interpreted as `float# c` instead of `float #c`. This is in theory a breaking change, but it doesn't break any code in opam or elsewhere.

Review: @ccasin?

### Implementation

The distinction between `float# c` and `float #c` is done in the lexer, where we produce a different token in the first case. This requires some extra attention to make infix hash operators, such as `#~#` in `x#~#y`, continue to lex properly.

### Surprising things

This is extremely specialized to `float#`: `int#`, `Float.t#`, etc. are parse errors.

### CODE REVIEW

File: ocaml/parsing/lexer.mll
Comment (ccasin) [inline ocaml/parsing/lexer.mll:155]: As discussed in person, I think the terminology "end_of_lexbuf" is a little confusing because there is lots more stuff in the lexbuf.  Possibly `enqueue_token_from_end_of_window` with a definition of window somewhere would be fine.  I also think `enqueue_token_at_start_of_lexbuf` would make sense (as this token is about to be the moral start of the lexbuf).

File: ocaml/parsing/lexer.mll
Comment (ccasin) [inline ocaml/parsing/lexer.mll:137]: I think this queue will only ever have 0 or 1 things in it.  It's fine to use a queue for the convenient interface, but maybe note that fact in one of your comments for clarity.

**Review by ccasin — APPROVED** [review]
This looks good.  I have a few minor suggestions in comments, but am pre-emptively approving.  Merge it when you are happy.

I considered suggesting setting up a nightly build to make sure the performance impact of querying the queue on every token isn't noticeable, but on reflection it's hard to imagine the cost of one additional branch per token will amount to anything.  Perhaps worth double checking in compiler UI after merging, though.

**Review by ccasin — APPROVED** [review]
This looks good.  I have a few minor suggestions in comments, but am pre-emptively approving.  Merge it when you are happy.

I considered suggesting setting up a nightly build to make sure the performance impact of querying the queue on every token isn't noticeable, but on reflection it's hard to imagine the cost of one additional branch per token will amount to anything.  Perhaps worth double checking in compiler UI after merging, though.

File: ocaml/testsuite/tests/typing-layouts/parsing_unboxed_floats.ml
Comment (ncik-roberts) [reply ocaml/testsuite/tests/typing-layouts/parsing_unboxed_floats.ml:13]: SGTM, done.

File: ocaml/testsuite/tests/typing-layouts/parsing_unboxed_floats.ml
Comment (ncik-roberts) [reply ocaml/testsuite/tests/typing-layouts/parsing_unboxed_floats.ml:8]: Yes, but it needs rewording to make any sense. I've done this.

**Comment by ncik-roberts** [comment]
Thanks for the review. I've responded to the suggestions and will merge this once it turns green.

**Comment by ncik-roberts** [comment]
I've pushed a little more code to improve an error message (that was made worse in a rebase). I'm confident the new code I added only touches a small corner of the Jane Syntax framework, but in the interest of avoiding surprises, @ccasin would you mind taking a quick look before I merge anything?

File: ocaml/testsuite/tests/typing-layouts/parsing_beta.compilers.reference
Comment (ccasin) [inline ocaml/testsuite/tests/typing-layouts/parsing_beta.compilers.reference:14]: I think this error is a little confusing (though the old one wasn't great either, so I'm not particularly blaming you).  It would be nice to say something like:
```
"float#" requires the "alpha" version of the "layouts" extension, which is disabled.
```
Maybe we could achieve this by having `Disabled_extension` carry an optional message prefix string so we can provide a little context-relevant information?

Comment (ncik-roberts) [reply ocaml/testsuite/tests/typing-layouts/parsing_beta.compilers.reference:14]: Your idea is nice. I don't want to see it all the way through right now, but I can at least make the layouts error message here a little more intelligible.

**Pull request merged**
